rpc: Add a `-rpcwaittimeout` parameter to limit time spent waiting

cdecker opened the pull request:
Adds a new numeric `-rpcwaittimeout` that can be used to limit the
time we spend waiting on the RPC server to appear. This is used by
downstream projects to provide a bit of slack when `bitcoind`s RPC
interface is not available right away.

This makes the `-rpcwait` argument more useful, since we can now limit
how long we'll ultimately wait, before potentially giving up and reporting
an error to the caller. It was discussed in the context of the BTCPayServer
wanting to have c-lightning wait for the RPC interface to become available
but still have the option of giving up eventually ([4355]).

I checked with laanwj whether this is already possible ([comment]), and
whether this would be a welcome change. Initially I intended to repurpose
the (optional) argument to `-rpcwait`, however I decided against it since it
would potentially break existing configurations, using things like `rpcwait=1`,
or `rpcwait=true` (the former would have an unintended short timeout, when
old behavior was to wait indefinitely).

~Due to its simplicity I didn't implement a test for it yet, but if that's desired I
can provide one.~ Test was added during reviews.

[4355]: https://github.com/ElementsProject/lightning/issues/4355
[comment]: https://github.com/ElementsProject/lightning/issues/4355#issuecomment-768288261

cdecker committed:
rpc: Add a `-rpcwaittimeout` parameter to limit time spent waiting

Adds a new numeric `-rpcwaittimeout` that can be used to limit the
time we spend waiting on the RPC server to appear. This is used by
downstream projects to provide a bit of slack when `bitcoind`s RPC
interface is not available right away.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -712,6 +714,9 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
     // Execute and handle connection failures with -rpcwait.
     const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
+    const int timeout = gArgs.GetArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT);
+    const int64_t deadline = GetTime() + timeout;
```
I think `GetTime` is deprecated in favor of `GetSystemTimeInSeconds` (not mockable) or `GetTime<T>` (mockable), see src/util.time.h

cdecker replied:
Ok, amended the use of `GetTime` with the `GetTime<T>` version (had to look it up, hopefully I got the right version).

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -70,6 +71,7 @@ static void SetupCliArgs(ArgsManager& argsman)
     argsman.AddArg("-rpcport=<port>", strprintf("Connect to JSON-RPC on <port> (default: %u, testnet: %u, signet: %u, regtest: %u)", defaultBaseParams->RPCPort(), testnetBaseParams->RPCPort(), signetBaseParams->RPCPort(), regtestBaseParams->RPCPort()), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::OPTIONS);
     argsman.AddArg("-rpcuser=<user>", "Username for JSON-RPC connections", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-rpcwait", "Wait for RPC server to start", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-rpcwaittimeout=<n>", strprintf("Timeout in seconds to wait for the RPC server to start, or 0 for no timeout. (default: %d)", DEFAULT_WAIT_CLIENT_TIMEOUT), ArgsManager::ALLOW_INT, OptionsCategory::OPTIONS);
```
Judging from the PR description you likely thought about this, but could the timeout still be an argument in `-rpcwait` and only kick in if the value is greater than 1? That would avoid introducing another config option and as a bonus would be shorter to type. IDK what kind of timeout values API clients need.

cdecker replied:
I'm quite happy to switch it around, but having `1` as a special value is kinda weird as well. I thought I'd go the most cautious route first and see which way wins in the review xD

jonatack replied:
Fair. Is the real time ever that low anyway? Maybe counting -rpcwait=1 as one second would be fine.

cdecker replied:
I don't expect many users to run into issues if we were to special case `-rpcwait=1`, but given the "creativity" of users I wouldn't put it past them to use `-rpcwait=42` or `-rpcwait=31337`. Some users seem to see "any value" as a personal challenge :smiley:

jonatack replied:
I'd still be in favor of combining `rpcwait` and `rpcwaittimeout`, but if we prefer two arguments for this, it may be helpful to mention the relationship to `-rpcwait` in this description.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -39,6 +39,7 @@ UrlDecodeFn* const URL_DECODE = urlDecode;
 
 static const char DEFAULT_RPCCONNECT[] = "127.0.0.1";
 static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;
+static const int DEFAULT_WAIT_CLIENT_TIMEOUT=0;
```
could be `constexpr`

cdecker replied:
Had to catch up about `constexpr`, but good point :+1:

jonatack reviewed (commented):
Concept ACK. Could append a test to the existing `-rpcwait` coverage at the end of `test/functional/interface_bitcoin_cli.py` to add coverage.

theStack commented:
Concept ACK

laanwj commented:
Concept ACK

practicalswift commented:
Concept ACK

cdecker commented:
Thanks for the reviews and concept ACKs everybody :+1: 

I'm a bit puzzled by the appveyo failure, which doesn't surface any details related to these changes, is this something my changes broke or was it something with `master` at the time of the branching?

promag commented:
Concept ACK, code change looks good to me.

Just noting that effective minimum timeout is 1s, maybe sleep with min(timout, 1s)? An alternative would be to specify attempt count.

cdecker commented:
> Just noting that effective minimum timeout is 1s, maybe sleep with min(timout, 1s)? An alternative would be to specify attempt count.

Excellent point/ The reason I went for the deadline approach was that in between sleeps we are actually attempting to connect over the network, which has its own timeout (`DEFAULT_HTTP_CLIENT_TIMEOUT`), which would cause drift if we were to just count the attempts instead of using the delta.

Effectively the time we can wait in the worst case is `-rpcwaittimeout + DEFAULT_HTTP_CLIENT_TIMEOUT` if we happen to start an attempt right before the deadline gets triggered. If we were counting we'd have a maximum of `-rpcwaittimeout * (1s + DEFAULT_HTTP_CLIENT_TIMEOUT)` , which isn't really the interface I'd feel comfortable using.

jonatack commented:
> I'm a bit puzzled by the appveyo failure, which doesn't surface any details related to these changes, is this something my changes broke or was it something with `master` at the time of the branching?

Appveyor is a bit flakey ATM, it's probably unrelated.

cdecker commented:
Rebased on top of `master`, let's see if the flake persists.

cdecker commented:
Yep seems, like CI is happier now ^^

promag reviewed (commented):
Code review ACK 7dcead4319258b058eb8ee5f2271e5c6dad264aa.

nit, could have a release note.

darosior commented on src/bitcoin-cli.cpp:
```diff
@@ -713,6 +715,9 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
     // Execute and handle connection failures with -rpcwait.
     const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
+    const int timeout = gArgs.GetArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT);
+    const int64_t deadline = GetTime<std::chrono::microseconds>().count() + timeout;
```
I think you want this to be `std::chrono::seconds` instead of `std::chrono::microseconds`, or to change the doc above to mention the parameter is passed in microseconds.

darosior reviewed (changes requested):
Concept ACK

Small note too: you should remove the `@` from the OP or laanwj will get spammed by all the altcoins merge (and the merge script will refuse it anyways) :)

cdecker commented:
> Code review ACK [7dcead4](https://github.com/bitcoin/bitcoin/commit/7dcead4319258b058eb8ee5f2271e5c6dad264aa).
> 
> nit, could have a release note.

Can you point me to the conventions for release notes? This is my first PR that might need them 🎉

laanwj commented:
> Can you point me to the conventions for release notes? This is my first PR that might need them 

I couldn't find anything in `CONTRIBUTING.md` (this would make sense to add). We prefer to use 'release notes fragments' to avoid introducing merge hotspots. These will be merged before the next major release.

These have a) the broad category (in this case, I guess "RPC client functionality") b) a very brief description (not full documentation!) of the user-visible change.

See for example: 
https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes-18077.md
https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes-19776.md

cdecker commented:
> > Can you point me to the conventions for release notes? This is my first PR that might need them
> 
> I couldn't find anything in `CONTRIBUTING.md` (this would make sense to add). We prefer to use 'release notes fragments' to avoid introducing merge hotspots. These will be collected before the next major release.
> 
> These have a) the broad category (in this case, I guess "RPC client functionality") b) a very brief description (not full documentation!) of the user-visible change.
> 
> See for example:
> https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes-18077.md
> https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes-19776.md

Awesome, thanks. I was expecting something like this to avoid hotspots, and it's way easier than c-lightning's Changelog-in-the-commit-message method :+1:

promag commented:
We have details at https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes

laanwj commented:
> We have details at https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes

Oh great! I somehow couldn't find this (I did grep for a few things). Might make sense to link to this from `CONTRIBUTING.md`.

cdecker committed:
doc: Add release notes for the `-rpcwaittimeout` cli parameter

cdecker commented:
Added preliminary release notes in a new file, and fixed up the units in the `GetTime<T>` template as well as the unit of the sleep call. Let me know if the release notes are not in the desired format, and I'll fix them up asap.

jonatack commented on doc/release-notes-21056.md:
```diff
@@ -0,0 +1,6 @@
+
+- The new `-rpcwaittimeout` argument to `bitcoin-cli` sets the timeout
+  in seconds to use with `-rpcwait`. If the timeout expires
+  `bitcoin-cli` will report a failure. (#21056)
```
Among the current headers in `doc/release-notes.md`, "New settings" might be the closest, but I'm not sure. There doesn't seem to be a CLI header. Suggestion follows, feel free to ignore.

```diff
-RPC client functionality
-------------------------
+New settings
+------------
 
-- The new `-rpcwaittimeout` argument to `bitcoin-cli` sets the timeout
-  in seconds to use with `-rpcwait`. If the timeout expires
+- A new `-rpcwaittimeout` argument to `bitcoin-cli` sets the timeout
+  in seconds to use with `-rpcwait`. If the timeout expires,
   `bitcoin-cli` will report a failure. (#21056)
```

laanwj replied:
I don't know if that header is better. As a compromise "New bitcoin-cli settings" maybe?

I do think it's important to keep mentions of server and other tools's command line arguments separated. Then again, all of this can be organized when merging the fragments—it's not an automatic process.

jonatack replied:
Yes, "New bitcoin-cli settings" seems better to me, too. Though as mentioned in https://github.com/bitcoin/bitcoin/pull/21056#discussion_r576917497, if we add the argument to the existing `-rpcwait=` (as a CLI setting intended for human manual use rather than by client software, ISTM it may be less subject to API constraints and would be less verbose than `-rpcwait -rpcwaittimeout=`), it could be "Updated bitcoin-cli settings".

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -757,8 +762,9 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
-            if (fWait) {
-                UninterruptibleSleep(std::chrono::milliseconds{1000});
+            const int64_t now = GetTime<std::chrono::seconds>().count();
+            if (fWait && (timeout == 0 || now < deadline)) {
```
The error I'm seeing is a bit confusing. It would be good to raise a specific error for the timeout value exceeded.
```
$ bitcoind -regtest -daemon ; bitcoin-cli -regtest -rpcwait -rpcwaittimeout=3 -getinfo
Bitcoin Core starting
error: server in warmup
```

jonatack reviewed (commented):
A few comments below. Haven't checked, but ISTM the logic could use `std::chrono` throughout rather than converting values back to integers. A functional test verifying the timeout works correctly and a specific error type is raised might be a good idea.

luke-jr reviewed (approved):
utACK, but I think actual software shouldn't be using bitcoin-cli at all...

laanwj commented:
> utACK, but I think actual software shouldn't be using bitcoin-cli at all...

For what it's worth I agree with @luke-jr here. Never really understood why `lightningd` doesn't use the RPC API directly. It's generally the more stable way, less moving parts, less potential ambiguity in input arguments. But it's an aside of course and off topic here.

cdecker commented:
> For what it's worth I agree with @luke-jr here. Never really understood why `lightningd` doesn't use the RPC API directly. It's generally the more stable way, less moving parts, less potential ambiguity in input arguments. But it's an aside of course and off topic here.

Not trying to make this a `lightningd` specific proposal, but since it was brought up I'll try to reconstruct the rationale at the time.

It was/is easier to call out to an existing program that includes the logic for authentication, error handling and timeouts than to implement it from scratch on top of out io library which doesn't have good http support (this is likely self-inflicted, but it's what we've got and it's working well for us so far). Authentication here being the big win, since if you have `bitcoin-cli` working you don't have to do anything to configure `lightningd` to be able to talk to `bitcoind` (no need to handle cookie auth, user/pass auth, binding and connecting, no need to implement TLS). If we had implemented all these things we'd likely have ended up replicating all the functionality `bitcoin-cli` already has, which didn't seem like a good investment of time given the already huge amount of work to be done.

Are there downsides of calling out to `bitcoin-cli`? Sure. Will we eventually end up replacing it with an internal implementation? Maybe. Is it working well as it is now? I think so 😉

That being said, I'm sure we're not the only ones that could use a wait timeout (thinking primarily of scripts and operations).

laanwj commented:
> Are there downsides of calling out to bitcoin-cli? Sure. Will we eventually end up replacing it with an internal implementation? Maybe. Is it working well as it is now? I think so

I guess it's mostly down to personal preference,  I prefer to use APIs directly, calling out to external programs reminds me of shell scripts which always are very environment-dependent. One concrete case is that bitcoin RPC might be forwarded from another container. Right now it's necessary to copy `bitcoin-cli` to the container hosting c-lightning. 

Also as `bitcoin-cli` is aimed at human users, it has some client-side processing for arguments—some arguments are parsed as JSON others string-ized. Sending JSON directly is more "pure". Due to this added "moving part" I would personally prefer to avoid it from robust software.

Also accessing the API directly gives you full control over things such as timeouts. No need to wait for PRs to be merged here and then part of a release so that users have it :smile:  JSONRPC handling is straightforward if you already have JSON for your own APIs.

> It was/is easier to call out to an existing program that includes the logic for authentication

For authentication—I'd somewhat agree, but we don't do anything complex such as HMAC or challenge/response. The only available authentication mechanism is HTTP basic authentication with a)  send `$DATADIR/.cookie` as user/password or b) manually provide user/password.

> no need to implement TLS

There hasn't been TLS support in `bitcoin-cli` for years.

Sorry for derailing this further. Again, this is not meant as criticism on this change nor on `lightningd` (and I'm not volunteering to make a patch to change this). As you say, it works  :woman_shrugging:

cdecker commented:
Reworded the release notes as proposed by @jonatack, and rebased on top of `master`.

Open questions are:

 - Add a specific error message pointing towards a timeout if we get something like `RPC warmup`
   - Pointed out [here](https://github.com/bitcoin/bitcoin/pull/21056#discussion_r576924237)
   - I think passing through the error is correct here, and not amend it since the timeout is merely retrying a failing attempt and eventually we give up passing through the cause. If we were to customize the error we could end up hiding the underlying cause for the failure (timeout ok, but what caused the timeout to fail ultimately?)
 - Instead of adding a new command line option in the form of `-rpcwaittimeout`, merge it instead with `-rpcwait`, redefining the semantics of the value. See above for discussion but here's a summary:
   - Pro: merging avoids adding a new option, making the expression more concise
   - Con: users may have scripts and utilities that use `-rpcwait=1` to opt into the waiting, but with the new semantics we'd just wait for 1 second. Also suddenly things like `-rpcwait=true` would be invalid or we'd require workarounds mapping `true`, `on` etc to `0`.

Anything I'm missing?

laanwj commented:
> If we were to customize the error we could end up hiding the underlying cause for the failure (timeout ok, but what caused the timeout to fail ultimately?)

Tend to agree. Maybe printing both would be a possibility? E.g. `timeout on transient error: …`.

> Instead of adding a new command line option in the form of -rpcwaittimeout, merge it instead with -rpcwait, redefining the semantics of the value.

I'm, in the abstract, not a big fan of Swiss-army-knife overloading of arguments to options either. It quickly gets too clever for its own good. Would be fine with keeping it like this.

cdecker commented:
> Tend to agree. Maybe printing both would be a possibility? E.g. `timeout on transient error: …`.
> 
> >

I'll look into it, sounds like a good idea :+1:

cdecker committed:
rpc: Prefix rpcwaittimeout error with details on its nature

As proposed by @laanwj the error message is now prefixed with the
"timeout on transient error:" prefix, to explain why the error is
suddenly considered terminal.

cdecker commented:
Prefixed the error message with the prefix proposed by @laanwj. One thing that I'm not quite clear on is whether the string allocated by `strprintf` is freed (not too familiar how the memory ownership in the presence of exceptions in C++ is). Would that require additional handling or is it ok in this case to wait for the exit to clean up memory?

laanwj commented:
Thanks!

> One thing that I'm not quite clear on is whether the string allocated by strprintf is freed 

Yes, the exception object (including the string it refers to) is destructed when it leaves the scope where it is catched. This should be okay.

cdecker committed:
rpc: Add test for -rpcwaittimeout

Suggested-by: Jon Atack <@jonatack>

cdecker commented:
> Concept ACK. Could append a test to the existing `-rpcwait` coverage at the end of `test/functional/interface_bitcoin_cli.py` to add coverage.

Going through the reviews I noticed that I hadn't addressed this so far. Added a test for it where suggested.

cdecker commented:
As far as I can see the only unresolved thing in this PR is the discussion about whether to reuse `-rpcwait` and redefine the argument to be the timeout in seconds, or whether the current addition of a new flag is preferrable.

promag commented on src/bitcoin-cli.cpp:
```diff
@@ -738,6 +740,9 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
     UniValue response(UniValue::VOBJ);
     // Execute and handle connection failures with -rpcwait.
     const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
+    const int timeout = gArgs.GetArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT);
```
nit, is it fine to accept <0? A negative value is same as  `-rpcwait=false`.

cdecker replied:
I think so, the alternative would be to throw an error, but that might be a bit much.

promag commented on src/bitcoin-cli.cpp:
```diff
@@ -748,11 +753,12 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
-                UninterruptibleSleep(std::chrono::milliseconds{1000});
+        } catch (const CConnectionFailed& e) {
+            const int64_t now = GetTime<std::chrono::seconds>().count();
+            if (fWait && (timeout == 0 || now < deadline)) {
```
Maybe `timeout <= 0` - zero or negative means no timeout.

cdecker replied:
Added check using `assert_greater_than_or_equal` and had to import `time` but that should work ok.

promag reviewed (commented):
Tested ACK 4840dee30d8c679a2df999d7b9d48e19653ad3bd.

> As far as I can see the only unresolved thing in this PR is the discussion about whether to reuse `-rpcwait` and redefine the argument to be the timeout in seconds, or whether the current addition of a new flag is preferrable.

I think current approach is fine otherwise how would the user disable waiting - `-rpcwait=0` is disable or wait forever?

@cdecker you can update OP now that there is a test.

cdecker commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -246,6 +246,10 @@ def run_test(self):
+        self.log.info("Test -rpcwait option waits at most -rpcwaittimeout seconds for startup")
+        self.stop_node(0)  # stop the node so we time out
+        assert_raises_process_error(1, "Could not connect to the server", self.nodes[0].cli('-rpcwait', '-rpcwaittimeout=5').echo)
+
```
Added check using `assert_greater_than_or_equal` and had to import `time` but that should work ok.

promag replied:
This should be `assert_greater_than_or_equal(time.time(), start_time + 5)`.

cdecker commented:
Rebased on top of `master`, addressed @promag's feedback and amended OP to mention the tests that are now included.

cdecker commented:
Sorry, used to the c-lightning workflow where we keep fixups to show how things were addressed. Will squash as soon as I wake up 🙂

cdecker commented:
Rebased on top of `master`

promag reviewed (commented):
Tested ACK 5685cdcacd but the test needs to be fixed.

promag reviewed (commented):
ACK b9e76f1bf08c52fcd402b2314e00db4ad247ebc8.

cdecker commented:
> Tested ACK [5685cdc](https://github.com/bitcoin/bitcoin/commit/5685cdcacdb8b59e420c4cc32a1ef3bf2cac067c) but the test needs to be fixed.

Yep, I switched the arguments around apparently :wink:

cdecker commented:
Is this good in its current state or does something need to be addressed? Probably not the highest priority issue, but would love to remove it from my watchlist, e.g., by merging it :wink:

laanwj commented:
Code review ACK b9e76f1bf08c52fcd402b2314e00db4ad247ebc8
